feat: add `blas/ext/base/ndarray/ssorthp`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Progresses #2656.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `blas/ext/base/ndarray/ssorthp`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add blas/ext/base/ndarray/ssorthp

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/benchmark/benchmark.js:
```diff
@@ -0,0 +1,105 @@
+// VARIABLES //
+
+var options = {
+	'dtype': 'generic'
```
```suggestion
	'dtype': 'float32'
```

kgryte committed:
bench: fix dtype

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/docs/repl.txt:
```diff
@@ -0,0 +1,41 @@
+
+    Examples
+    --------
+    > var xbuf = [ 1.0, -2.0, 3.0, -4.0 ];
```
```suggestion
    > var xbuf = new {{alias:@stdlib/array/float32}}( [ 1.0, -2.0, 3.0, -4.0 ] );
```

kgryte committed:
docs: fix example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/docs/repl.txt:
```diff
@@ -0,0 +1,41 @@
+    > {{alias}}( [ x, o ] )
+    <ndarray>
+    > var data = x.data
+    [ -4.0, -2.0, 1.0, 3.0 ]
```
```suggestion
    <Float32Array>[ -4.0, -2.0, 1.0, 3.0 ]
```

kgryte committed:
docs: fix return value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/docs/types/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+* var ndarray = require( '@stdlib/ndarray/base/ctor' );
+*
+* var xbuf = new Float32Array( [ 1.0, -2.0, 3.0, -4.0 ] );
+* var x = new ndarray( 'generic', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
```
```suggestion
* var x = new ndarray( 'float32', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
```

kgryte committed:
docs: fix dtype

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/docs/types/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+* var arr = ndarra2array( out );
+* // returns [ -4.0, -2.0, 1.0, 3.0 ]
+*/
+declare function ssorthp<T = unknown>( arrays: [ typedndarray<T>, typedndarray<number> ] ): typedndarray<T>;
```
This isn't correct. The input ndarray needs to be a `float32ndarray`.

kgryte replied:
And similarly, the output ndarray should be a `float32ndarray`. You'll need to update `test.ts`, as well.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/docs/types/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+
+/**
+* Sorts a one-dimensional single-precision floating-point ndarray using heapsort.
+*
```
I believe you are missing the sort order note here.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/lib/index.js:
```diff
@@ -0,0 +1,53 @@
+* var ssorthp = require( '@stdlib/blas/ext/base/ndarray/ssorthp' );
+*
+* var xbuf = new Float32Array( [ 1.0, -2.0, 3.0, -4.0 ] );
+* var x = new ndarray( 'generic', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
```
```suggestion
* var x = new ndarray( 'float32', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
```

kgryte committed:
docs: fix dtype

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/docs/types/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+* var out = ssorthp( [ x, ord ] );
+* // returns <ndarray>
+*
+* var arr = ndarra2array( out );
```
```suggestion
* var arr = ndarray2array( out );
```

kgryte committed:
docs: fix function name

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/lib/index.js:
```diff
@@ -0,0 +1,53 @@
+* var out = ssorthp( [ x, ord ] );
+* // returns <ndarray>
+*
+* var arr = ndarra2array( out );
```
```suggestion
* var arr = ndarray2array( out );
```

kgryte committed:
docs: fix function name

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+* var ndarray = require( '@stdlib/ndarray/base/ctor' );
+*
+* var xbuf = new Float32Array( [ 1.0, -2.0, 3.0, -4.0 ] );
+* var x = new ndarray( 'generic', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
```
```suggestion
* var x = new ndarray( 'float32', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
```

kgryte committed:
docs: fix dtype

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/ssorthp/test/test.js:
```diff
@@ -0,0 +1,76 @@
+	});
+
+	actual = ssorthp( [ x, order ] );
+	t.strictEqual( actual.dtype, x.dtype, 'returns expected value' );
+	t.deepEqual( [ 1.0, 2.0, 3.0, 4.0, 5.0 ], x.data, 'returns expected value' );
+	t.deepEqual( actual.shape, x.shape, 'returns expected value' );
+	t.deepEqual( actual.strides, x.strides, 'return expected value' );
+	t.strictEqual( actual.offset, x.offset, 'returns expected value' );
```
You need to test against actual values. Since the function mutates the input array, comparing to `x.dtype`, `x.shape`, etc doesn't actually test anything. Those assertions will always be true, no matter if the function does incorrect things.

Also, we should be using functions for accessing ndarray attributes. E.g., `getDType( actual )`, `getData( actual )`, etc.

```suggestion
	t.strictEqual( actual, x, 'returns expected value' );
	t.strictEqual( actual.dtype, 'float32', 'returns expected value' );
	t.deepEqual( actual.data, new Float32Array( 1.0, 2.0, 3.0, 4.0, 5.0 ] ), 'returns expected value' );
	t.deepEqual( actual.shape, [ 5 ], 'returns expected value' );
	t.deepEqual( actual.strides, [ 1 ], 'return expected value' );
	t.strictEqual( actual.offset, 0, 'returns expected value' );
```

Applies here and below.

May also require updating the tests for `gsorthp`.

kgryte reviewed (changes requested):
Left an initial review.

headlessNode committed:
fix: apply suggestion from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---
